### PR TITLE
get interface name dynamically

### PR DESCRIPTION
**Pull request by StefanoSalsano**
removed if_name_for_flannel from defaults.yaml
(obviously flannel was not working in the libvirt testbed because the interface was different... and we needed to change it manually in env.yaml)
now the variable if_name_for_flannel is set by ansible in kubernetes.yaml, by executing a shell script on the node
note that the ip command is not working in this shell script, I have replaced with 
/usr/sbin/ip which is working on both:
    virtualbox: "bento/centos-7.4"
    libvirt: "centos/7"
we should investigate why /usr/sbin is not in the path of the shell launched by ansible, while it is ok when ansible executes configure-ipv6-interface.sh

### CODE REVIEW

**Comment by ferrarimarco** [comment]
My guess is that the two tasks are using different shells.

In `configure-ipv6-interface.sh` we explicitly ask for `#!/bin/sh` (and by the way, it should be `#!/usr/bin/env sh` because different distributions put the POSIX shell (`dash`, or other shells (`bash`, `zsh`) in `sh mode` in different places), while in the shell task you just implemented uses the default shell, which, in this case is `bash`.

Shell initialization is not a trivial thing:

shell startup flow.

I bet that `/usr/sbin` is being added to the `PATH` by `bash`-specific configuration files.

That's why I implemented a [modular approach in my dotfiles](https://github.com/ferrarimarco/dotfiles).

**Comment by StefanoSalsano** [comment]
@ferrarimarco according to https://unix.stackexchange.com/questions/197488/centos-7-shell-script-interpretor, #!/bin/sh is actually bash on centos
I think that also #!/usr/bin/env sh points to bash on centos, and on dash on ubuntu
I am going now to investigate more on this problem, so that we can discuss a solution...
first I will try to verify which shell is called by inline scripts in ansible

**Comment by ferrarimarco** [comment]
If you ask for `#!/usr/bin/env sh` you get a POSIX-compliant shell (dash on Ubuntu, or Bash in sh mode, as you mentioned). For us it doesn't actually matter, as long as it's POSIX-compliant.

POSIX-compliance is not an hard requirement, but a best practice to have more portable scripts.

As you can see from the [Ansible docs](https://docs.ansible.com/ansible/latest/modules/shell_module.html), the shell module uses `/bin/sh`. This looks a bit strange to me, because if you use bashisms (such as pipefail) it doesn't complain. I guess it's using the user's default shell, where user = vagrant in our case.

File: Vagrantfile [hunk 1]
```diff
@@ -137,8 +137,6 @@ kubernetes_minion_1_ipv6 = network_prefix_ipv6 + settings["net"]["minion_1_ipv6_
 kubernetes_minion_2_ipv6 = network_prefix_ipv6 + settings["net"]["minion_2_ipv6_part"]
```
Comment (ferrarimarco) [inline Vagrantfile:137]: Reminder: Add an echo that shows the currently selected provider.

**Comment by StefanoSalsano** [comment]
- [X] removed if_name_for_flannel from defaults.yaml

- [X] Add an echo that shows the currently selected provider  and networking plugin

- [X] Add a provisioner `--provision-with ansible-debug`  that only executes tasks tagged with `ansible_debug` 

NB we discovered that to run the ip command used to discover the interface name on the nodes, we need to give the root permission (enable: true) to the task

File: Vagrantfile (hunk 1 above)
Comment (StefanoSalsano) [reply Vagrantfile:142]: done

**Comment by StefanoSalsano** [comment]
> 
> 
> If you ask for `#!/usr/bin/env sh` you get a POSIX-compliant shell (dash on Ubuntu, or Bash in sh mode, as you mentioned). For us it doesn't actually matter, as long as it's POSIX-compliant.
> 
> POSIX-compliance is not an hard requirement, but a best practice to have more portable scripts.
> 
> As you can see from the [Ansible docs](https://docs.ansible.com/ansible/latest/modules/shell_module.html), the shell module uses `/bin/sh`. This looks a bit strange to me, because if you use bashisms (such as pipefail) it doesn't complain. I guess it's using the user's default shell, where user = vagrant in our case.

ok, we have clarified that it's the other way around, with `/bin/sh` we get a posix compliant shell, with  `#!/usr/bin/env sh` you get exactly sh

File: ansible/kubernetes.yml [hunk 2]
```diff
@@ -9,12 +9,26 @@
       listen:
         - systemd_read_configs
   post_tasks:
+    - name: ansible_debug
+        This task is executed by vagrant provision --provision-with ansible-debug
```
Comment (ferrarimarco) [inline ansible/kubernetes.yml:13]: Are you sure this format works?

Btw, I'd rephrase this to avoid mentioning vagrant, because it's already in the readme, and to avoid passive forms. Apart from that, this is a great help

File: ansible/kubernetes.yml [hunk 3]
```diff
@@ -9,12 +9,26 @@
...
+        this task is only an example, it can be replace by your own debug task
+        or other tasks with ansible_debug tag can be created
+      become: true
+      changed_when: false
+      shell: |
+        set -e ; \
+        echo $0 ; \
+        whoami ; \
+        echo $PATH
+      register: ansible_debug_output
```
Comment (ferrarimarco) [inline ansible/kubernetes.yml:23]: Then, what are you doing with this variable? Are you writing it somewhere as output?

**Comment by ferrarimarco** [comment]
> ok, we have clarified that it's the other way around, with `/bin/sh` we get a posix compliant shell, with `#!/usr/bin/env sh` you get exactly sh

What do you mean by "exactly sh"? In CentOS for example, /bin/sh is symlinked to /bin/bash as you said, so there's no "sh", but just bash in sh mode. But it doesn't matter, we're switching to Debian/Ubuntu anyway. :)

**Comment by StefanoSalsano** [comment]
@ferrarimarco on centos, in particular I'm using bento/centos-7.4, with sh you get sh, with /bin/sh you get bash... for this reason in scripts/linux/enable-ssh-password-authentication.sh we should change #!/usr/bin/env sh and use #!/bin/sh as we have in other scripts

see the output from my VM hereafter:
```
[vagrant@k8s-master-1 ~]$ which sh
/usr/bin/sh
```
so if you run  /usr/bin/env sh you get
```
[vagrant@k8s-master-1 ~]$ /usr/bin/env sh
sh-4.2$ echo $0
sh
```

File: ansible/kubernetes.yml (hunk 3 above)
Comment (StefanoSalsano) [reply ansible/kubernetes.yml:23]: @ferrarimarco this is an example, but you can read the variable in the -vv log, so in principle you do not need to print (this is how I've used it)...
we can add a printout for the clarity of the example (any suggestion?) or I can add the instructions to read the "-vv" log in the task description

**Comment by ferrarimarco** [comment]
I'll do some checks but from what you say, on CentOS we have:

```shell
[vagrant@k8s-master-1 ~]$ which sh
/usr/bin/sh
```

```shell
[vagrant@k8s-master-1 ~]$ /usr/bin/env sh
sh-4.2$ echo $0
sh
```

We should check:

1. `ls -al /usr/bin/sh`
1. `ls -al /bin/sh`
1. `echo "$(/usr/bin/env sh)"`

PS: I'm ok in changing `#!/usr/bin/env sh` to `#!/bin/sh` in `scripts/linux/enable-ssh-password-authentication.sh` as you suggest. This is discussion is more to understand how things are working, but we can move this conversation to another channel so we can merge this PR after you change the shebang on that script :)

Anyway, thanks!

File: ansible/kubernetes.yml (hunk 3 above)
Comment (ferrarimarco) [reply ansible/kubernetes.yml:23]: Right! I forgot about the fact that we add -vv sometimes.

I was thinking more about this: https://docs.ansible.com/ansible/latest/modules/debug_module.html

File: Vagrantfile [hunk 4]
```diff
@@ -97,6 +97,11 @@ if File.exist?(env_specific_config_path)
   end
 end
 
+# Display the main current configuration parameters
+@ui.info "Welcome to Kubernetes playground !"
```
Comment (ferrarimarco) [inline Vagrantfile:101]: Extra space after `playground`. Remove

File: Vagrantfile [hunk 5]
```diff
@@ -97,6 +97,11 @@ if File.exist?(env_specific_config_path)
...
+@ui.info "Vagrant provider : " + settings["conf"]["vagrant_provider"]
```
Comment (ferrarimarco) [inline Vagrantfile:102]: 1) `Currently selected Vagrant provider:`
2) I think we defined and initialized a `vagrant_provider` variable somewhere in the `Vagrantfile`, we should use that.

File: Vagrantfile [hunk 6]
```diff
@@ -97,6 +97,11 @@ if File.exist?(env_specific_config_path)
...
+@ui.info "Networking plugin : " + settings["ansible"]["group_vars"]["all"]["kubernetes_network_plugin"]
```
Comment (ferrarimarco) [inline Vagrantfile:103]: 1) `Currently selected Kubernetes networking plugin:`
2) Same thing about the variable. Did we define one?

File: ansible/kubernetes.yml (hunk 3 above)
Comment (StefanoSalsano) [reply ansible/kubernetes.yml:23]: ok I've added a second debug task to msg the output of the first one :-)

File: Vagrantfile (hunk 5 above)
Comment (StefanoSalsano) [reply Vagrantfile:102]: it's defined later... I've considered this issue, I have a small preference to output here exactly the raw values that are in the setting, and this is the logical place where to output it... even before we do the checks... 
yes, there is a cost of reading twice some parameters from settings but i would pay it so that we have a clean and simple approach: when we want to output some value to the ui, we just go in this place, we read and write the value
do you have a strong preference for the other approach?

File: Vagrantfile (hunk 6 above)
Comment (StefanoSalsano) [reply Vagrantfile:103]: see above

Comment (StefanoSalsano) [reply Vagrantfile:103]: BTW in this case we do not have a variable... so it would be case by case...

Comment (ferrarimarco) [reply Vagrantfile:103]: Ok, let's not lose time on this one

File: Vagrantfile (hunk 5 above)
Comment (ferrarimarco) [reply Vagrantfile:102]: I got what you say.

What if we output the whole configuration in a compact form? With a trick, you can have it as a nicely formatted YAML `settings.to_yaml` and print that, so we take the raw values as they are, and we don't have to bother about adding more stuff to this output when we add parameters.

Note that the `.to_yaml` method outputs, well, YAML. This might clutter things a bit, so we might have to find a more compact representation.

File: Vagrantfile (hunk 4 above)
Comment (StefanoSalsano) [reply Vagrantfile]: I added on purpose because graphically was more evident...
De gustibus non disputandum ... I'll leave the ui of k8s-pg to you :-)
I've removed it
I'll open an issue to propose:
1. Removal of the message about vagrant version
2. Adding the k8s-pg version number in the Welcome line

File: README.md [hunk 7]
```diff
@@ -208,10 +208,26 @@ You can also run the same test suite locally. To bootstrap a development
 environment, you need to install the runtime dependencies listed above, plus the
 development environment dependencies.
 
-### Debug output
+### Debugging ansible operations
 
 Ansible output is saved in the `/vagrant/ansible_output.txt`.
 
+For debbugging and development purpose, you can run a Vagrant provisioner called
```
Comment (ferrarimarco) [inline README.md:215]: `purposes`

File: Vagrantfile (hunk 5 above)
Comment (StefanoSalsano) [reply Vagrantfile:102]: IMO the full configuration file is overkill for the normal usage... but this is my opinion...
in this case I prefer to have three lines that gives me the important information
we can open a issue to discuss further

File: README.md (hunk 7 above)
Comment (StefanoSalsano) [reply README.md]: done

File: ansible/kubernetes.yml [hunk 8]
```diff
@@ -9,12 +9,35 @@
       listen:
         - systemd_read_configs
   post_tasks:
+    - name: ansible_debug_example_1
+        To execute this task, call
```
Comment (ferrarimarco) [inline ansible/kubernetes.yml:13]: I think my previous comment slipped through :(

I'd remove mentions to vagrant from the description of this task, because Ansible has no notion of Vagrant. Additionally, we say the same thing in the README, so no need to specify it here.
